nifi-nar-bundles/nifi-spark-bundle/nifi-livy-controller-service/src/main/java/org/apache/nifi/controller/livy/LivyBatchController.java

```diff
@@ -0,0 +1,279 @@
+        headers.put("Content-Type", LivySessionService.APPLICATION_JSON);
+        headers.put("X-Requested-By", LivySessionService.USER);
+
+        log.debug("submitBatch() Submitting Batch Job to Spark via: " + batchUrl);
+
+        StringBuilder payload = new StringBuilder("{\"file\":\"" + codePath + "\"");
+        appendValue(payload, "className", className);
+
+        try {
+            if (codeArgs != null) {
+                payload.append(",\"args\":");
+                payload.append(splitCollect(codeArgs));
+            }
+            if (jars != null) {
+                payload.append(",\"jars\":");
+                payload.append(splitCollect(jars));
+            }
+            if (files != null) {
+                payload.append(",\"files\":");
+                payload.append(splitCollect(files));
+            }
+            if (pyFiles != null) {
+                payload.append(",\"pyFiles\":");
+                payload.append(splitCollect(pyFiles));
+            }
+            if (archives != null) {
+                payload.append(",\"archives\":");
+                payload.append(splitCollect(archives));
+            }
+        } catch (IOException e){
+            throw new SessionManagerException(e);
+        }
+
+        appendValue(payload, "proxyUser", proxyUser);
+        appendValue(payload, "queue", queue);
+        appendValue(payload, "driverMemory", driverMemory);
+        appendValue(payload, "driverCores", driverCores);
+        appendValue(payload, "executorMemory", executorMemory);
+        appendValue(payload, "executorCores", executorCores);
+        appendValue(payload, "numExecutors", numExecutors);
+        appendSubObject(payload, "conf", conf);
+
+        payload.append("}");
+
+        log.debug("submitBatch() Payload: " + payload.toString());
+
+        try {
+            output = LivyHelpers.readJSONObjectFromUrlPOST(
+                    sslContextService, credentialsService, connectTimeout, batchUrl, headers, payload.toString());
+
+            return output.getInt("id");
+        } catch (JSONException|IOException e){
+            throw new SessionManagerException(e);
+        }
+    }
+
+    final ObjectMapper mapper = new ObjectMapper();
+
+    private void appendSubObject(StringBuilder payload, String key, String val){
+        if (val != null) {
+            payload.append(",\"" + key + "\": ");
+            payload.append(val);
+        }
+    }
+
+    private void appendValue(StringBuilder payload, String key, String val){
+        if (val != null) {
+            payload.append(",\"" + key + "\": \"");
+            payload.append(val);
+            payload.append("\"");
+        }
+    }
+
+    private void appendValue(StringBuilder payload, String key, Integer val){
+        if (val != null) {
+            payload.append(",\"" + key + "\": ");
+            payload.append(val);
+        }
+    }
+
+    private String splitCollect(String arg) throws IOException {
```

Comment:
It would be nicer to rename `transformCsvToJsonArrayStr`.

Reply:
Thanks for the feedback, but it is supposed to do two things :)
In the UI, NiFi users provide a comma delimited list of values. In the code, we accept that comma delimited string and need to convert it to JSON. This method handles that.